{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "monthly-offer",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# Quantifying methylation data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "supreme-fighter",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Methods overview"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rough-robinson",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## Input\n",
    "1. `phenoFile`: The input of this module is a folder containing all the IDAT file, 1 pair for each sample. Directly under the folder, there should be one companion csv file that documenting all the meta-information of the bisulfite sequencing. Please specify the `phenoFile` as the path to this companion csv file\n",
    "\n",
    "2. `cross_reactive`: A list of cpg probe that are reported to [map to multiple regions in the genome.](https://academic.oup.com/nargab/article/2/4/lqaa105/6040968) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "former-desert",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Output\n",
    "\n",
    "A pair bed.gz file for both beta and m value.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "olympic-optics",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Minimal working example\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "temporal-worse",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "## Command interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "supported-mozambique",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run RNA_calling.ipynb [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  fastqc\n",
      "  fastp_trim_adaptor\n",
      "  trimmomatic\n",
      "  STAR_align\n",
      "  rnaseqc_call\n",
      "  rsem_call\n",
      "  picard_qc\n",
      "\n",
      "Global Workflow Options:\n",
      "  --cwd output (as path)\n",
      "                        The output directory for generated files.\n",
      "  --samples VAL (as path, required)\n",
      "                        Sample meta data list\n",
      "  --data-dir  path(f\"{samples:d}\")\n",
      "\n",
      "                        Raw data directory, default to the same directory as\n",
      "                        sample list\n",
      "  --job-size 1 (as int)\n",
      "                        For cluster jobs, number commands to run per job\n",
      "  --walltime 5h\n",
      "                        Wall clock time expected\n",
      "  --mem 16G\n",
      "                        Memory expected\n",
      "  --java-mem 6G\n",
      "                        Memory for Java virtual mechine (`picard`)\n",
      "  --numThreads 8 (as int)\n",
      "                        Number of threads\n",
      "  --container ''\n",
      "                        Software container option\n",
      "  --gtf VAL (as path, required)\n",
      "                        Reference gene model\n",
      "\n",
      "Sections\n",
      "  fastqc:\n",
      "  fastp_trim_adaptor:\n",
      "    Workflow Options:\n",
      "      --window-size 4 (as int)\n",
      "                        sliding window setting\n",
      "      --required-quality 20 (as int)\n",
      "      --leading 20 (as int)\n",
      "                        the mean quality requirement option for cut_front\n",
      "      --trailing 20 (as int)\n",
      "                        the mean quality requirement option for cut_tail\n",
      "      --min-len 15 (as int)\n",
      "                        reads shorter than length_required will be discarded\n",
      "  trimmomatic:\n",
      "    Workflow Options:\n",
      "      --trimmomatic-jar '/opt/Trimmomatic-0.39/trimmomatic-0.39.jar'\n",
      "                        Path to the software. Default set to using our\n",
      "                        rna_quantification.sif image\n",
      "      --fasta-with-adapters-etc VAL (as str, required)\n",
      "                        Illumina clip setting Path to the reference adaptors\n",
      "      --seed-mismatches 2 (as int)\n",
      "      --palindrome-clip-threshold 30 (as int)\n",
      "      --simple-clip-threshold 10 (as int)\n",
      "      --window-size 4 (as int)\n",
      "                        sliding window setting\n",
      "      --required-quality 20 (as int)\n",
      "      --leading 3 (as int)\n",
      "                        Other settings\n",
      "      --trailing 3 (as int)\n",
      "      --min-len 50 (as int)\n",
      "  STAR_align:\n",
      "    Workflow Options:\n",
      "      --STAR-index VAL (as path, required)\n",
      "                        STAR indexing file\n",
      "      --outFilterMultimapNmax 20 (as int)\n",
      "                        Alignment parameter\n",
      "      --alignSJoverhangMin 8 (as int)\n",
      "      --alignSJDBoverhangMin 1 (as int)\n",
      "      --outFilterMismatchNmax 999 (as int)\n",
      "      --outFilterMismatchNoverLmax 0.1 (as float)\n",
      "      --alignIntronMin 20 (as int)\n",
      "      --alignIntronMax 1000000 (as int)\n",
      "      --alignMatesGapMax 1000000 (as int)\n",
      "      --outFilterType BySJout\n",
      "      --outFilterScoreMinOverLread 0.33 (as float)\n",
      "      --outFilterMatchNminOverLread 0.33 (as float)\n",
      "      --limitSjdbInsertNsj 1200000 (as int)\n",
      "      --outSAMstrandField intronMotif\n",
      "      --outFilterIntronMotifs None\n",
      "      --alignSoftClipAtReferenceEnds Yes\n",
      "      --quantMode TranscriptomeSAM GeneCounts (as list)\n",
      "      --outSAMattrRGline ID:rg1 SM:sm1 (as list)\n",
      "      --outSAMattributes NH HI AS nM NM ch (as list)\n",
      "      --chimSegmentMin 15 (as int)\n",
      "      --chimJunctionOverhangMin 15 (as int)\n",
      "      --chimOutType Junctions WithinBAM SoftClip (as list)\n",
      "      --chimMainSegmentMultNmax 1 (as int)\n",
      "  rnaseqc_call_1, rsem_call_1, picard_qc:\n",
      "    Workflow Options:\n",
      "      --ref-flat . (as path)\n",
      "                        Path to flat reference file, for computing QC metric\n",
      "      --picard-jar '/opt/picard-tools/picard.jar'\n",
      "                        Path to the software. Default set to using our\n",
      "                        rna_quantification.sif image\n",
      "      --reference-fasta VAL (as path, required)\n",
      "                        The fasta reference file used to generate star index\n",
      "      --optical-distance 100 (as int)\n",
      "                        For the patterned flowcell models (HiSeq X), change to\n",
      "                        2500\n",
      "  rnaseqc_call_2:\n",
      "    Workflow Options:\n",
      "      --strand VAL (as str, required)\n",
      "                        Options are fr or rf or unstranded\n",
      "  rnaseqc_call_3:\n",
      "  rsem_call_2:\n",
      "    Workflow Options:\n",
      "      --RSEM-index VAL (as path, required)\n",
      "      --max-frag-len 1000 (as int)\n",
      "      --strand VAL (as str, required)\n",
      "                        Options are fr or rf or unstranded\n",
      "  rsem_call_3:\n",
      "  rsem_call_4, rnaseqc_call_4:\n",
      "  rsem_call_5, rnaseqc_call_5:\n",
      "    Workflow Options:\n",
      "      --ref-flat . (as path)\n",
      "                        Path to flat reference file, for computing QC metric\n"
     ]
    }
   ],
   "source": [
    "sos run methylation_calling.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "alpine-frame",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Setup and global parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "passing-measurement",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# The output directory for generated files.\n",
    "parameter: cwd = path(\"output\")\n",
    "# The companion csv file as outlined in the input section.\n",
    "parameter: phenoFile = path\n",
    "# Raw data directory, has to be the folder containing the companion csv file as outlined in the input section.\n",
    "data_dir = path(f\"{phenoFile:d}\")\n",
    "# For cluster jobs, number commands to run per job\n",
    "parameter: job_size = 1\n",
    "# Wall clock time expected\n",
    "parameter: walltime = \"5h\"\n",
    "# Memory expected\n",
    "parameter: mem = \"16G\"\n",
    "\n",
    "# Number of threads\n",
    "parameter: numThreads = 8\n",
    "# Software container option\n",
    "parameter: container = \"\"\n",
    "from sos.utils import expand_size\n",
    "cwd = path(f'{cwd:a}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "actual-commission",
   "metadata": {
    "kernel": "SoS",
    "toc-hr-collapsed": true
   },
   "source": [
    "## Step 1: \n",
    "The first step of methylation processing is to acquired the beta and M value after preliminary QC and filtering of the IDAT data.\n",
    "\n",
    "By default the data will be annotated based on hg38 using [this annotation](https://github.com/achilleasNP/IlluminaHumanMethylationEPICanno.ilm10b5.hg38), alternatively user can set the `--hg` parameter back to 19 to use the [hg19 annotation](https://bioconductor.org/packages/release/data/annotation/html/IlluminaHumanMethylationEPICanno.ilm10b4.hg19.html) \n",
    "\n",
    "1. Based on 1 input csv file, all the IDAT file in the folder and sub-folder will be loaded\n",
    "2. The methylation data samples will first be filtered based on [bisulphite conversation rate](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4527772/) \n",
    "\n",
    "3. samples will then be filtered based on a [detection pvalue](https://www.rdocumentation.org/packages/minfi/versions/1.18.4/topics/detectionP), which indicates the quality of the signal at each genomics position\n",
    "4. Quantile Normalization will then be applied\n",
    "5. features will be filtered if they are on sex chr, known to be [cross-reactive,maping to multiple regions in the genome.](https://academic.oup.com/nargab/article/2/4/lqaa105/6040968), overlapping with snps, or having too low a detection P\n",
    "6. Beta and M value will for all the probes/samples will then each be saved to a TSV file.\n",
    "\n",
    "\n",
    "[As documented here](https://github.com/cumc/xqtl-pipeline/issues/312) when the batch of IDAT data are different, there will be a problem reading the IDAT file without specifing the force = TRUE option in the `read.metharray.exp(targets = targets,force = TRUE)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "refined-bidder",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[methylation_1]\n",
    "parameter: samples_pval_tre = 0.05\n",
    "parameter: probe_det_pval_tre = 0.01\n",
    "parameter: bisulfite_conversion_tre = 85\n",
    "parameter: cross_reactive = path(\".\")\n",
    "parameter: hg = 38 #hg = 38 or 19, by default 38\n",
    "input: samples\n",
    "output: f'{cwd}/{name}.minifi.rds',f'{cwd}/{name}.methyl.beta.bed',f'{cwd}/{name}.methyl.M.bed',f'{cwd}/{name}.methyl.M.region_list',f'{cwd}/{name}.methyl.sample_list'  \n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads\n",
    "R: expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout', container=container\n",
    "    ## load libraries\n",
    "    library(minfi)\n",
    "    library(dplyr)\n",
    "    library(tibble)\n",
    "    library(IlluminaHumanMethylationEPICmanifest)\n",
    "    if (${cross_reactive} != \".\"){cross_reactive = attach(${cross_reactive}) } else {cross_reactive = NULL}\n",
    "    \n",
    "    ## Define functions\n",
    "    bscon_minfi <- function(RGsetEx){\n",
    "        \n",
    "        #getting the values from the green channel\n",
    "        (csp.green <- function (RGsetEx, controls = c(\"BISULFITE CONVERSION I\", \"BISULFITE CONVERSION II\"))\n",
    "        {\n",
    "          minfi:::.isRGOrStop(RGsetEx)\n",
    "          r <- getRed(RGsetEx)\n",
    "          g <- getGreen(RGsetEx)\n",
    "          sapply (controls, function( controlType ) {\n",
    "        \n",
    "           ctrlAddress <- try (getControlAddress(RGsetEx, controlType = controlType), silent = T)\n",
    "           if (!inherits (ctrlAddress, 'try-error')){ctrlAddress <- getControlAddress(RGsetEx, controlType = controlType)}\n",
    "           else\n",
    "             stop (\"450k QC data could not be found\")\n",
    "        \n",
    "        \n",
    "           g[ctrlAddress, ]\n",
    "          })})\n",
    "        \n",
    "        green <- csp.green(RGsetEx)\n",
    "        \n",
    "        #Getting values from the red channel\n",
    "        (csp.red <- function (RGsetEx, controls = c(\"BISULFITE CONVERSION I\", \"BISULFITE CONVERSION II\"))\n",
    "        {\n",
    "          minfi:::.isRGOrStop(RGsetEx)\n",
    "          r <- getRed(RGsetEx)\n",
    "          g <- getGreen(RGsetEx)\n",
    "          sapply (controls, function( controlType ) {\n",
    "        \n",
    "            ctrlAddress <- getControlAddress(RGsetEx, controlType = controlType)\n",
    "        \n",
    "            r[ctrlAddress, ]\n",
    "          })})\n",
    "        \n",
    "        red <- csp.red (RGsetEx)\n",
    "        \n",
    "        #selecting only the Bisulfite conversion I values from both green and red\n",
    "        bsI.green <- green$`BISULFITE CONVERSION I`\n",
    "        bsI.red <- red$`BISULFITE CONVERSION I`\n",
    "        #selecting only the Bisulfite conversion II values from both green and red\n",
    "        bsII.green <- green$`BISULFITE CONVERSION II`\n",
    "        bsII.red <- red$`BISULFITE CONVERSION II`\n",
    "        \n",
    "        # calculate BS conv type I betas as an example of using an index vector\n",
    "        if(nrow(bsI.green) > 11){ # 450K\n",
    "          BSI.betas <- rbind(bsI.green[1:3,], bsI.red[7:9,])/((rbind(bsI.green[1:3,], bsI.red[7:9,])) + rbind(bsI.green[4:6,], bsI.red[10:12,]))\n",
    "        } else { # EPIC\n",
    "          BSI.betas <- rbind(bsI.green[1:2,], bsI.red[6:7,])/((rbind(bsI.green[1:2,], bsI.red[6:7,])) + rbind(bsI.green[3:4,], bsI.red[ 8:9 ,]))\n",
    "        }\n",
    "        \n",
    "        #calculation of BS con in Type II data\n",
    "        BSII.betas <- bsII.red/(bsII.red + bsII.green)\n",
    "        \n",
    "        apply(rbind(BSI.betas, BSII.betas), 2, median)*100 ## this is the value you are interested in\n",
    "        }\n",
    "\n",
    "    ## 1. read idat files\n",
    "    targets <- read.metharray.sheet(${_input:d})\n",
    "    \n",
    "    rgSet <- read.metharray.exp(targets = targets)\n",
    "    if(${hg} == 38){rgSet@annotation = c(array = \"IlluminaHumanMethylationEPIC\", annotation = \"ilm10b5.hg38\")}\n",
    "    message(\"rgset created\")\n",
    "    \n",
    "    ###### Quality Control and Normalization ###############\n",
    "    ## 2. bisulphite conversation rate filtering\n",
    "    rgSet_bcr = bscon_minfi(rgSet)\n",
    "    rgSet = rgSet[,names(which(rgSet_bcr > ${bisulfite_conversion_tre} ))]\n",
    "    ## 3. QC based on p-value, remove samples with average p value less than 0.05\n",
    "    \n",
    "    detP <- detectionP(rgSet)\n",
    "    keep <- colMeans(detP) < ${samples_pval_tre}\n",
    "    rgSet <- rgSet[,keep]\n",
    "    targets <- targets[keep,]\n",
    "    message(\"samples with avg det p-val < ${samples_pval_tre} removed\")\n",
    "\n",
    "    ## 4. Normalize the data - Quantile\n",
    "    mSetSq <- preprocessQuantile(rgSet)\n",
    "    message(\"data quantile-normalized\")\n",
    "    ## 5. Remove XY chr - probes\n",
    "    \n",
    "    mSetSq <- mSetSq[!as.vector(mSetSq@rowRanges@seqnames)%in% c(\"chrX\",\"chrY\"),]\n",
    "    message(\"sex probes removed\")\n",
    "    \n",
    "    ## 6. Remove cross-reactive probes\n",
    "    no_cross_reactive <- !(featureNames(mSetSq) %in% cross_reactive)\n",
    "    mSetSq <- mSetSq[no_cross_reactive, ]\n",
    "    message(\"cross-reactive probes removed\")\n",
    "    \n",
    "    ## 7. Drop probes that are also SNPs\n",
    "    \n",
    "    mSetSq <- dropLociWithSnps(mSetSq)\n",
    "    message(\"probes overlapping with snps removed\")\n",
    "    \n",
    "    ## 8. Remove probes with < ${probe_det_pval_tre} detectin p-values\n",
    "    detP <- detP[match(featureNames(mSetSq),rownames(detP)),]\n",
    "    keep <- rowSums(detP < ${probe_det_pval_tre}) == ncol(mSetSq))\n",
    "    mSetSq <- mSetSq[keep,]\n",
    "    \n",
    "    ## 9. get Beta and Mvalues\n",
    "    mSetSqbval <- getBeta(mSetSq)%>%as_tibble(rownames = \"ID\")\n",
    "    mSetSqMval <- getM(mSetSq)%>%as_tibble(rownames = \"ID\")\n",
    "    cpg_regions = mSetSq@rowRanges%>%as.data.frame()%>%as_tibble(rownames = \"ID\")%>%select(\"#chr\" = seqnames, start, end, ID, -width,-strand)\n",
    "    mSetSqbval = cpg_regions%>%right_join(mSetSqbval, by = \"ID\")%>%mutate(end = start + 1)\n",
    "    mSetSqMval = cpg_regions%>%right_join(mSetSqMval, by = \"ID\")%>%mutate(end = start + 1)\n",
    "    message(\"Beta-values and M-values obtained\")\n",
    "    ## 10. output data\n",
    "    mSetSqMval = mSetSqMval%>%rename_at(vars(rgSet@colData%>%rownames()), function(x) rgSet@colData[x,]%>%as_tibble%>%pull(Sample_Name) )\n",
    "    mSetSqMval[,1:4]%>%write_delim(${_output[3]:r},\"\\t\")\n",
    "    mSetSqbval%>%readr::write_delim(\"${_output[1]}\",\"\\t\")\n",
    "    mSetSqMval%>%readr::write_delim(\"${_output[2]}\",\"\\t\")\n",
    "    output = list(\"rgSet\" = rgSet,mSetSq = \"mSetSq\", mSetSqbval = \"mSetSqbval\", mSetSqMval = \"mSetSqMval\")\n",
    "    output%>%saveRDS(${_output[0]})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "929fa247-1b74-4b08-a155-94226293ad1f",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Step 2\n",
    "After the beta and m value are generated. The output will be saved as a bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37af90fa-9316-4ef9-99e7-bb248bffdb98",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[methylation_2]\n",
    "output: f'{cwd}/{name}.methyl.beta.bed.gz',f'{cwd}/{name}.methyl.M.bed.gz' \n",
    "bash: expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout', container=container\n",
    "    bgzip -f ${_input[1]} \n",
    "    tabix ${_output[0]}\n",
    "    bgzip -f ${_input[2]} \n",
    "    tabix ${_output[1]}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "956092ba-89db-4d65-a1c0-ce8ba2459abc",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Archived: combat batch effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33d5766c-21a9-4514-89f7-aafc526bbfb5",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[combat_batch]\n",
    "parameter: batchFile = path(\".\")\n",
    "input: output_from(\"methylation\")\n",
    "output:f'{_input}.MDS.pdf',f'{_input}.beta.batch_corrected.txt'\n",
    "R: expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout', container=container\n",
    "    library(ggplot2)\n",
    "    library(metaMA)\n",
    "    library(sva)\n",
    "    library(lme4)\n",
    "    ## Define functions\n",
    "    posibatches <- function(dat, Sentrix, batch=TRUE, par.prior=TRUE, prior.plots=FALSE, mean.only.posi=FALSE, mean.only.batch=FALSE) {\n",
    "      require(sva)\n",
    "      require(lme4)\n",
    "      #get the position and batch information----------------------------------------\n",
    "      ### Extraction of the position numbers and chip numbers--------------------\n",
    "      #if (is.null(SentrixVector)){\n",
    "      if (is.null(Sentrix)){\n",
    "        stop('Sentrix informations must be provided.')\n",
    "      }\n",
    "      #chips<-as.numeric(factor(substr(Sentrix$sampleNames, 1, 10))) \n",
    "      #positions<-as.numeric(factor(substr(Sentrix$sampleNames, 12, 17))) \n",
    "      \n",
    "      chips <- as.numeric(factor(sapply(strsplit(as.character(Sentrix$sampleNames), \"_\"), function(x) x[1])))\n",
    "      positions <- as.numeric(factor(sapply(strsplit(as.character(Sentrix$sampleNames), \"_\"), function(x) x[2])))\n",
    "      \n",
    "      if (length(positions)!=length(chips)){\n",
    "        stop('positions and chips must have the same length')\n",
    "      }\n",
    "      if (sum(positions>12)>0){\n",
    "        stop('Position number cannot be greater than 12')\n",
    "      }\n",
    "      if (sum(is.na(chips))>1 || sum(is.na(positions))>1){\n",
    "        stop('One or more position or chip numbers missing')\n",
    "      }\n",
    "      \n",
    "      if (sum(is.na(Sentrix$batches))==length(chips)){\n",
    "        batches<-chips\n",
    "      } else {\n",
    "        batches<-as.factor(Sentrix$batches)\n",
    "      }\n",
    "      \n",
    "      if(batch==TRUE){\n",
    "        \n",
    "        ################################################################################\n",
    "        ################################################################################\n",
    "        pct_threshold = .8 # Amount of variability desired to be explained by the principal components.  Set to match the results in book chapter and SAS code.  User can adjust this to a higher (>= 0.8) number but < 1.0\n",
    "        dataRowN <- nrow(dat)\n",
    "        dataColN <- ncol(dat)\n",
    "        \n",
    "        ########## Center the data (center rows) ##########\n",
    "        datCentered <- matrix(data = 0, nrow = dataRowN, ncol = dataColN)\n",
    "        datCentered_transposed = apply(dat, 1, scale, center = TRUE, scale = FALSE)\n",
    "        datCentered = t(datCentered_transposed)\n",
    "        \n",
    "        exp_design<-data.frame(cbind(positions,batches))\n",
    "        expDesignRowN <- nrow(exp_design)\n",
    "        expDesignColN <- ncol(exp_design)\n",
    "        myColNames <- names(exp_design)\n",
    "        \n",
    "        \n",
    "        ########## Compute correlation matrix ##########\n",
    "        \n",
    "        theDataCor <- cor(datCentered)\n",
    "        \n",
    "        ########## Obtain eigenvalues ##########\n",
    "        \n",
    "        eigenData <- eigen(theDataCor)\n",
    "        eigenValues = eigenData$values\n",
    "        ev_n <- length(eigenValues)\n",
    "        eigenVectorsMatrix = eigenData$vectors\n",
    "        eigenValuesSum = sum(eigenValues)\n",
    "        percents_PCs = eigenValues /eigenValuesSum\n",
    "        \n",
    "        ########## Merge experimental file and eigenvectors for n components ##########\n",
    "        \n",
    "        my_counter_2 = 0\n",
    "        my_sum_2 = 1\n",
    "        for (i in ev_n:1){\n",
    "          my_sum_2  = my_sum_2 - percents_PCs[i]\n",
    "          if ((my_sum_2) <= pct_threshold ){\n",
    "            my_counter_2 = my_counter_2 + 1\n",
    "          }\n",
    "          \n",
    "        }\n",
    "        if (my_counter_2 < 3){\n",
    "          pc_n  = 3\n",
    "          \n",
    "        }else {\n",
    "          pc_n = my_counter_2\n",
    "        }\n",
    "        \n",
    "        # pc_n is the number of principal components to model\n",
    "        \n",
    "        pc_data_matrix <- matrix(data = 0, nrow = (expDesignRowN*pc_n), ncol = 1)\n",
    "        mycounter = 0\n",
    "        for (i in 1:pc_n){\n",
    "          for (j in 1:expDesignRowN){\n",
    "            mycounter <- mycounter + 1\n",
    "            pc_data_matrix[mycounter,1] = eigenVectorsMatrix[j,i]\n",
    "            \n",
    "          }\n",
    "        }\n",
    "        \n",
    "        AAA <- exp_design[rep(1:expDesignRowN,pc_n),]\n",
    "        \n",
    "        Data <- cbind(AAA,pc_data_matrix)\n",
    "        \n",
    "        ####### Edit these variables according to your factors #######\n",
    "        \n",
    "        variables <- c(colnames(exp_design))\n",
    "        for (i in 1:length(variables)) {\n",
    "          Data$variables[i] <- as.factor(Data$variables[i])\n",
    "        }\n",
    "        \n",
    "        ########## Mixed linear model ##########\n",
    "        \n",
    "        p <- options(warn = (-1))\n",
    "        #effects_n = expDesignColN + choose(expDesignColN, 2) + 1\n",
    "        effects_n = expDesignColN  + 1\n",
    "        randomEffectsMatrix <- matrix(data = 0, nrow = pc_n, ncol = effects_n)\n",
    "        \n",
    "        model.func <- c()\n",
    "        index <- 1\n",
    "        for (i in 1:length(variables)) {\n",
    "          mod = paste(\"(1|\", variables[i], \")\", sep = \"\")\n",
    "          model.func[index] = mod\n",
    "          index = index + 1\n",
    "        }\n",
    "        \n",
    "        function.mods <- paste(model.func, collapse = \" + \")\n",
    "        \n",
    "        for (i in 1:pc_n) {\n",
    "          y = (((i - 1) * expDesignRowN) + 1)\n",
    "          funct <- paste(\"pc_data_matrix\", function.mods, sep = \" ~ \")\n",
    "          Rm1ML <- lmer(funct, Data[y:(((i - 1) * expDesignRowN) +\n",
    "                                         expDesignRowN), ], REML = TRUE, control=lmerControl(check.nobs.vs.nlev = \"ignore\",check.nobs.vs.rankZ = \"ignore\",check.nobs.vs.nRE=\"ignore\"),verbose = FALSE,\n",
    "                        na.action = na.omit)\n",
    "          randomEffects <- Rm1ML\n",
    "          randomEffectsMatrix[i, ] <- c(unlist(VarCorr(Rm1ML)),\n",
    "                                        resid = sigma(Rm1ML)^2)\n",
    "        }\n",
    "        effectsNames <- c(names(getME(Rm1ML, \"cnms\")), \"resid\")\n",
    "        ########## Standardize Variance ##########\n",
    "        \n",
    "        randomEffectsMatrixStdze <- matrix(data = 0, nrow = pc_n, ncol = effects_n)\n",
    "        for (i in 1:pc_n){\n",
    "          mySum = sum(randomEffectsMatrix[i,])\n",
    "          for (j in 1:effects_n){\n",
    "            randomEffectsMatrixStdze[i,j] = randomEffectsMatrix[i,j]/mySum\n",
    "          }\n",
    "        }\n",
    "        \n",
    "        ########## Compute Weighted Proportions ##########\n",
    "        \n",
    "        randomEffectsMatrixWtProp <- matrix(data = 0, nrow = pc_n, ncol = effects_n)\n",
    "        for (i in 1:pc_n){\n",
    "          weight = eigenValues[i]/eigenValuesSum\n",
    "          for (j in 1:effects_n){\n",
    "            randomEffectsMatrixWtProp[i,j] = randomEffectsMatrixStdze[i,j]*weight\n",
    "          }\n",
    "        }\n",
    "        ######### Compute Weighted Ave Proportions ##########\n",
    "        \n",
    "        randomEffectsSums <- matrix(data = 0, nrow = 1, ncol = effects_n)\n",
    "        randomEffectsSums <-colSums(randomEffectsMatrixWtProp)\n",
    "        totalSum = sum(randomEffectsSums)\n",
    "        randomEffectsMatrixWtAveProp <- matrix(data = 0, nrow = 1, ncol = effects_n)\n",
    "        \n",
    "        for (j in 1:effects_n){\n",
    "          randomEffectsMatrixWtAveProp[j] = randomEffectsSums[j]/totalSum\n",
    "          \n",
    "        }\n",
    "        \n",
    "        if(randomEffectsMatrixWtAveProp[,1]<randomEffectsMatrixWtAveProp[,2]){\n",
    "          afterbatchExp<-ComBat(dat = dat, batch = batches, par.prior = par.prior, prior.plots = prior.plots, mean.only = mean.only.batch)\n",
    "          afterposiExp<-ComBat(dat = afterbatchExp, batch = positions, par.prior = par.prior, prior.plots = prior.plots, mean.only = mean.only.posi)\n",
    "          dat<-afterposiExp\n",
    "        } else{\n",
    "          afterposiExp<-ComBat(dat = dat, batch = positions, par.prior = par.prior, prior.plots = prior.plots, mean.only = mean.only.posi)\n",
    "          afterbatchExp<-ComBat(dat = afterposiExp, batch = batches, par.prior = par.prior, prior.plots = prior.plots, mean.only = mean.only.batch)\n",
    "          dat<-afterbatchExp\n",
    "        }\n",
    "      }else{\n",
    "        afterposiExp<-ComBat(dat = dat, batch = positions, par.prior = par.prior, prior.plots = prior.plots, mean.only = mean.only.posi)\n",
    "        dat<-afterposiExp\n",
    "      }\n",
    "      return(dat)\n",
    "    }\n",
    "    ## Combat batch effect\n",
    "    mSetSq = readRDS(\"${_input[0]}\")$mSetSq\n",
    "    Sentrix = read.csv(\"${batchFile}\")\n",
    "    results <- posibatches(mSetSq, Sentrix, batch=TRUE, par.prior=TRUE, prior.plots=FALSE, mean.only.posi=FALSE, mean.only.batch=FALSE)\n",
    "    results%>%readr::write_delim(${_output[1]:r})\n",
    "    ## Plot MDS\n",
    "    data4mds <- function(data, topvals){\n",
    "      b <- data\n",
    "      o <- order(rowVars(b), decreasing = TRUE)[seq_len(topvals)]\n",
    "      d <- dist(t(b[o, ]))\n",
    "      fit <- cmdscale(d)\n",
    "      return(fit)\n",
    "    }\n",
    "    \n",
    "    myfit <- as.data.frame(data4mds(mSetSq, 1000))\n",
    "    colnames(myfit) <- c(\"fit1\", \"fit2\")\n",
    "\n",
    "    MDSplot1 = ggplot(myfit, aes(x = fit1, y = fit2)) + geom_point(size = 1.5) + \n",
    "      labs(title = \"MDS plot\",  x = \"Coordinate 1\",  y = \"Coordinate 2\") + \n",
    "      theme(axis.title.x = element_text(face = \"bold\", size = 26), axis.text.x = element_text(face=\"bold\", size = 20)) + theme(axis.text.y = element_text(face=\"bold\", size = 20), axis.title.y = element_text(face = \"bold\", size = 26)) +  \n",
    "      theme(panel.background = element_rect(fill = \"white\", colour = \"white\")) + \n",
    "      theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(size = 0.5, linetype = \"solid\", colour = \"black\")) +  \n",
    "      theme(plot.title = element_text(face = \"bold\", size = 30, hjust = 0.5)) + \n",
    "      theme(legend.position=\"bottom\", legend.text=element_text(size=16))\n",
    "  MDSplot1%>%ggsave(${_output[0]:r})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     "shell"
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.22.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
