{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "sustainable-ability",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# cisQTL analysis workflows\n",
    "\n",
    "This section documents output from the Assocation Scan section of command generator MWE and explained the purpose for each of the command. The file used in this page can be found at [here](https://drive.google.com/drive/folders/16ZUsciZHqCeeEWwZQR46Hvh5OtS8lFtA?usp=sharing).\n",
    "\n",
    "There are two options availble: TensorQTL and APEX. For a gene-SNP pairs, they produce the same cis norminal association p-value.\n",
    "\n",
    "The default cis window is defined as 1 million dp up and down stream of the transcription start site(TSS). \n",
    "\n",
    "If bed file that were not generated by our workflow are used, please be noted that the end column should be TSS + 1 instead of the actual end of gene region.\n",
    "\n",
    "**Each commands in the Molecular Phenotype Processing tutorials will be generated once per theme. The MWE is considered a one theme analysis** \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "452dee1e-ca15-43ca-ab39-63f29b1f9e49",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## TensorQTL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36814462-59ad-47d1-a688-153f6589ff71",
   "metadata": {},
   "source": [
    "[TensorQTL](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1836-7) is a new implementation of [Matrix QTL](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3348564/). It will produce two gene-level pvalue for the association: one estimated using beta-distribution and one estimated using permutation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "394ebf32-dd0a-4dec-b862-aabb3f61cde7",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run pipeline/TensorQTL.ipynb cis \\\n",
    "    --genotype-list plink_files_list.txt   \\\n",
    "    --phenotype-list output/data_preprocessing/MWE/phenotype_data/MWE.log2cpm.bed.processed_phenotype.per_chrom.recipe \\\n",
    "    --covariate-file output/data_preprocessing/MWE/covariates/MWE.log2cpm.MWE.covariate.cov.MWE.MWE.related.filtered.extracted.pca.projected.resid.bed.PEER.MWE.covariate.cov.MWE.MWE.related.filtered.extracted.pca.projected.cov.gz    \\\n",
    "    --cwd output/association_scan/MWE/TensorQTL  \\\n",
    "    --container containers//TensorQTL.sif"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "reduced-static",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## APEX"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5259bcf8-1a91-4590-8dd1-734fe1811902",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "[APEX](https://www.biorxiv.org/content/10.1101/2020.12.18.423490v1) can use either a linear model or a linear mixed model for association testing. The linear mix model is potentially useful for analysis with related individuals. By default, our workflow use the linear model for better comparability with the TensorQTL method.\n",
    "\n",
    "For gene-level p-value, APEX use a recently-developed [aggregated Cauchy association test (ACAT) to combine p-values under arbitrary dependence structures](https://arxiv.org/abs/1808.09011) and should be comparable to beta-approximated permutation p-values. \n",
    "\n",
    "It should be noted that APEX requires indexed vcf.gz file as input, which was transformed from the per chromosome plink binary input(bed/bim/fam) for tensorQTL in our pipeline.\n",
    "\n",
    "There are various cautions in using the APEX, please see the [module page](https://github.com/cumc/xqtl-pipeline/blob/main/code/association_scan/APEX/APEX.ipynb) for detail."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "universal-chorus",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run pipeline/APEX.ipynb cis \\\n",
    "    --genotype-list vcf_files_list.txt   \\\n",
    "    --phenotype-list output/data_preprocessing/MWE/phenotype_data/MWE.log2cpm.bed.processed_phenotype.per_chrom.recipe \\\n",
    "    --covariate-file output/data_preprocessing/MWE/covariates/MWE.log2cpm.MWE.covariate.cov.MWE.MWE.related.filtered.extracted.pca.projected.resid.bed.PEER.MWE.covariate.cov.MWE.MWE.related.filtered.extracted.pca.projected.cov.gz    \\\n",
    "    --cwd output/association_scan/MWE/APEX  \\\n",
    "    --container containers//APEX.sif"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "Markdown",
     "markdown",
     "markdown",
     "",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.22.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
