{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "59d5b415-b698-4f4c-830f-fa4e63f0115f",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# SuSIE results post process"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c3e81ce-3d9e-43ba-9722-4a84e6a86936",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "This notebook is to post-process the susie results into different text file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3819b186-187f-4b00-b066-41cb774f5993",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "import glob\n",
    "import pandas as pd\n",
    "# Input\n",
    "parameter: rds_path = paths\n",
    "# For complete susie, region_list or tad_list, for susie_rss , LD region list \n",
    "parameter: region_list = path\n",
    "\n",
    "# A region list file documenting the chr_pos_ref_alt of a susie_object\n",
    "parameter: cwd = path(\"output\")\n",
    "parameter: name = \"demo\"\n",
    "region_tbl = pd.read_csv(region_list,sep = \"\\t\")\n",
    "\n",
    "## Path to work directory where output locates\n",
    "## Containers that contains the necessary packages\n",
    "parameter: container = \"\"\n",
    "# For cluster jobs, number commands to run per job\n",
    "parameter: job_size = 1\n",
    "# Wall clock time expected\n",
    "parameter: walltime = \"5h\"\n",
    "# Memory expected\n",
    "parameter: mem = \"16G\"\n",
    "# Number of threads\n",
    "parameter: numThreads = 20\n",
    "\n",
    "# use this function to edit memory string for PLINK input\n",
    "from sos.utils import expand_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e39ee42-5012-400d-9c53-982a5ea3087a",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[susie_to_tsv]\n",
    "input: rds_path, group_by = 1\n",
    "output: f\"{cwd}/{_input:bn}.lbf.tsv\",f\"{cwd}/{_input:bn}.variant.tsv\",f\"{cwd}/{_input:bn}.effect.tsv\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container\n",
    "    library(\"dplyr\")\n",
    "    library(\"tibble\")\n",
    "    library(\"purrr\")\n",
    "    library(\"readr\")\n",
    "    library(\"stringr\")\n",
    "    \n",
    "    extract_lbf = function(susie_obj){\n",
    "    \n",
    "    if(\"variants\" %in% names(susie_obj) ){\n",
    "    ss_bf = tibble(snps = susie_obj$variants%>%str_replace(\"_\",\":\"), cs_index = ifelse(is.null(susie_obj$sets$cs_index), 0, paste0(susie_obj$sets$cs_index,collapse =\",\")),names = \"LD1\" )\n",
    "    }\n",
    "      else \n",
    "      {\n",
    "    ss_bf = tibble(snps = susie_obj$variable_name, cs_index = ifelse(is.null(susie_obj$sets$cs_index), 0, paste0(susie_obj$sets$cs_index,collapse =\",\")),names = ${f'{_input:bnnn}'})\n",
    "     }\n",
    "    \n",
    "    ss_bf = ss_bf%>%cbind(susie_obj$lbf_variable%>%t)%>%as_tibble()\n",
    "    \n",
    "    return(ss_bf)\n",
    "    }\n",
    "    \n",
    "    extract_variants_pip = function(susie_obj,region_list){\n",
    "    susie_tb = tibble( variants =  names(susie_obj$pip)[which( susie_obj$pip >= 0)], \n",
    "                           snps_index = which(( susie_obj$pip >= 0))) %>% \n",
    "        mutate(chromosome = map_chr(variants, ~read.table(text = .x, sep = \":\")$V1%>%str_replace(\"chr\",\"\") ),\n",
    "                position  = map_chr(variants, ~read.table(text = .x, sep = \":\")$V2  ),\n",
    "                ref = map_chr(position , ~read.table(text = .x, sep = \"_\",colClasses = \"character\")$V2  ),\n",
    "                alt = map_chr(position , ~read.table(text = .x, sep = \"_\",colClasses = \"character\")$V3  ),\n",
    "                position  = map_dbl(position , ~read.table(text = .x, sep = \"_\",as.is = T)$V1  )\n",
    "                             )%>% mutate(cs_order =map(snps_index , ~tryCatch(which(pmap(list( a= susie_obj$sets$cs) , function(a) .x %in% a )%>%unlist()), error = function(e) return(F) )  )%>%as.numeric%>%replace_na(0),\n",
    "                             cs_id = map_chr(cs_order,~ifelse(.x ==0, \"None\" ,names(susie_obj$sets$cs)[.x] ) ),\n",
    "                             log10_base_factor = map_chr(snps_index,~paste0( susie_obj$lbf_variable[,.x],  collapse = \";\")),\n",
    "                             pip = map_dbl(snps_index,~( susie_obj$pip[.x])),\n",
    "                             posterior_mean = map_dbl(snps_index,~( susie_obj$coef[.x])),\n",
    "                             posterior_sd = map_dbl(snps_index,~(susie_get_posterior_sd(susie_obj)[.x])),\n",
    "                             z = posterior_mean/posterior_sd)\n",
    "  \n",
    "          susie_tb =  susie_tb%>%mutate(  molecular_trait_id = region_list$molecular_trait_id,\n",
    "                             finemapped_region_start = region_list$finemapped_region_start,\n",
    "                             finemapped_region_end = region_list$finemapped_region_end)\n",
    "              }\n",
    "          return(susie_tb)\n",
    "        }\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "\n",
    "     extract_effect_pip = function(susie_obj,region_list){\n",
    "      susie_effect_tb_ls = region_list%>%mutate(result_tb = list( tibble(phenotype = susie_obj$name,\n",
    "        V = susie_obj$V,effect_id = paste0(\"L\",1:length(V) ) ,\n",
    "        cs_log10bf = susie_obj$lbf,\n",
    "        cs_min_r2  = (susie_obj$sets$purity[effect_id,1])%>%as.numeric%>%replace_na(0),\n",
    "        cs_avg_r2  = (susie_obj$sets$purity[effect_id,2])%>%as.numeric%>%replace_na(0),\n",
    "        coverage = map(effect_id, ~susie_obj$sets$coverage[which(names(susie_obj$sets$cs) == .x )])%>%as.numeric%>%replace_na(0),\n",
    "        cs = map_chr(susie_obj$sets$cs[effect_id],~susie_tb$variants[.x]%>%paste0(collapse = \";\"))) ))%>%unnest(result_tb)\n",
    "        susie_effect_tb_ls$cs[susie_effect_tb_ls$cs == \"\"] = \"None\"\n",
    "      return(susie_effect_tb_ls)\n",
    "      }\n",
    "      } \n",
    "  \n",
    "    \n",
    "  \n",
    "  \n",
    "    susie_obj = readRDS(rds_path)\n",
    "    lbf = extract_lbf(susie_obj)\n",
    "    region_list = read_delim(\"${region_list}\",\"\\t\")\n",
    "    if(ncol(region_list) == 3 ){   region_list =  region_list%>%mutate(`#chr` = `#chr`%>%str_remove_all(\" \") , ID = paste0(`#chr`,\"_\",start,\"_\",stop) ) } # LD_list \n",
    "    if(region_list$start[1] - region_list$end[1]  == -1 ){ \n",
    "        region_list = region_list%>%mutate( start = start - ${windows} ,end = start +${windows}) # region_list for fix cis windows  \n",
    "          } \n",
    "      if(\"gene_id\" %in% colnames(region_list)){region_list = region_list%>%mutate(ID = gene_id)  } # region_list for gene\n",
    "    region_list = region_list%>%select(molecular_trait_id = ID, chromosome  = `#chr`,finemapped_region_start = start ,finemapped_region_end = end)  # Formatting\n",
    "    variants_pip = extract_variants_pip( susie_obj , region_list)\n",
    "    effect_pip = extract_effect_pip( susie_obj , region_list)\n",
    "    lbf%>%write_delim(\"${_output[0]}\",\"\\t\")\n",
    "    variants_pip%>%write_delim(\"${_output[1]}\",\"\\t\")\n",
    "    effect_pip%>%write_delim(\"${_output[2]}\",\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b50c78bb-4357-4146-be2a-445147f32213",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[susie_to_tsv]\n",
    "parameter: region_list = path()\n",
    "parameter: windows = 1000000\n",
    "parameter: cwd = path(\"output\")\n",
    "input: rds_path, group_by = 1\n",
    "# example input: demo.ENSG00000197375.unisusie.fit.ALL.rds\n",
    "output: f\"{cwd}/{_input:bn}.effect.tsv\",f\"{cwd}/{_input:bn}.variant.tsv\",f\"{cwd}/{_input:bn}.lbf.tsv\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container\n",
    "    library(\"susieR\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"tibble\")\n",
    "    library(\"purrr\")\n",
    "    library(\"readr\")\n",
    "    library(\"tidyr\")\n",
    "    library(\"stringr\")\n",
    "    \n",
    "    extract_lbf = function(susie_obj){\n",
    "    \n",
    "    if(\"variants\" %in% names(susie_obj) ){\n",
    "    ss_bf = tibble(snps = susie_obj$variants%>%str_replace(\"_\",\":\"), cs_index = ifelse(is.null(susie_obj$sets$cs_index), 0, paste0(susie_obj$sets$cs_index,collapse =\",\")),names = \"LD1\" )\n",
    "    }\n",
    "      else \n",
    "      {\n",
    "    ss_bf = tibble(snps = susie_obj$variable_name, cs_index = ifelse(is.null(susie_obj$sets$cs_index), 0, paste0(susie_obj$sets$cs_index,collapse =\",\")),names = ${_name})\n",
    "     }\n",
    "    \n",
    "    ss_bf = ss_bf%>%cbind(susie_obj$lbf_variable%>%t)%>%as_tibble()\n",
    "    \n",
    "    return(ss_bf)\n",
    "    }\n",
    "    \n",
    "    \n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "    # Get list of cs snps\n",
    "    rds_path = ${_input:r}\n",
    "    molecular_trait_id = ${_name}\n",
    "    tad_list = read_delim(\"${region_list}\",\"\\t\")\n",
    "    tad_list = tad_list%>%filter(gene_id == molecular_trait_id )\n",
    "    if(\"tad_index\" %in% colnames(tad_list) ){\n",
    "          tad_list = tad_list%>%select(molecular_trait_id = gene_id, chromosome  = `#chr`, finemapped_region_start = start_x ,finemapped_region_end = end_x)\n",
    "                                       } else {\n",
    "          tad_list = tad_list%>%mutate(\"chromosome\" =  `#chr`,\n",
    "                              molecular_trait_id = gene_id , \n",
    "                              finemapped_region_end = start+${windows}+1 )  \n",
    "        }\n",
    "    susie_tb_ls = list()\n",
    "    susie_effect_tb_ls = list()\n",
    "    susie_list = readRDS(rds_path)\n",
    "    \n",
    "    if(!is.null(susie_list$sets$cs)){\n",
    "        ## Get variant table \n",
    "        susie_tb = tibble( variants =  names(susie_list$pip)[which( susie_list$pip >= 0)], \n",
    "                           snps_index = which(( susie_list$pip >= 0))) %>% \n",
    "        mutate(chromosome = map_chr(variants, ~read.table(text = .x, sep = \":\")$V1%>%str_replace(\"chr\",\"\") ),\n",
    "                position  = map_chr(variants, ~read.table(text = .x, sep = \":\")$V2  ),\n",
    "                ref = map_chr(position , ~read.table(text = .x, sep = \"_\",colClasses = \"character\")$V2  ),\n",
    "                alt = map_chr(position , ~read.table(text = .x, sep = \"_\",colClasses = \"character\")$V3  ),\n",
    "                position  = map_dbl(position , ~read.table(text = .x, sep = \"_\",as.is = T)$V1  )\n",
    "                             )\n",
    "           \n",
    "        susie_tb_ls = susie_tb %>% mutate(cs_order =map(snps_index , ~tryCatch(which(pmap(list( a= susie_list$sets$cs) , function(a) .x %in% a )%>%unlist()), error = function(e) return(F) )  )%>%as.numeric%>%replace_na(0),\n",
    "                             cs_id = map_chr(cs_order,~ifelse(.x ==0, \"None\" ,names(susie_list$sets$cs)[.x] ) ),\n",
    "                             log10_base_factor = map_chr(snps_index,~paste0( susie_list$lbf_variable[,.x],  collapse = \";\")),\n",
    "                             pip = map_dbl(snps_index,~( susie_list$pip[.x])),\n",
    "                             posterior_mean = map_dbl(snps_index,~( susie_list$coef[.x])),\n",
    "                             posterior_sd = map_dbl(snps_index,~(susie_get_posterior_sd(susie_list)[.x])),\n",
    "                             z = posterior_mean/posterior_sd)%>%\n",
    "                    mutate(  molecular_trait_id = tad_list$molecular_trait_id,\n",
    "                             finemapped_region_start = tad_list$finemapped_region_start,\n",
    "                             finemapped_region_end = tad_list$finemapped_region_end)\n",
    "        write_delim(susie_tb_ls,paste0(\"${_output[1]:nnnn}.\",susie_list$name,\".variant.tsv\"),\"\\t\")                                                                     \n",
    "                                                                                                                                                                           \n",
    "       ## Get effect table\n",
    "      ## I make use of susie_list$name to catch the tissue name. This attribute is added in the previous step. But\n",
    "      ## it's not in all susie objects. Maybe we can use other strategy? eg. extract it from file name\n",
    "                                                                                     \n",
    "        susie_effect_tb_ls = tad_list%>%mutate(result_tb = list( tibble(phenotype = susie_list$name,\n",
    "        V = susie_list$V,effect_id = paste0(\"L\",1:length(V) ) ,\n",
    "        cs_log10bf = susie_list$lbf,\n",
    "        cs_min_r2  = (susie_list$sets$purity[effect_id,1])%>%as.numeric%>%replace_na(0),\n",
    "        cs_avg_r2  = (susie_list$sets$purity[effect_id,2])%>%as.numeric%>%replace_na(0),\n",
    "        coverage = map(effect_id, ~susie_list$sets$coverage[which(names(susie_list$sets$cs) == .x )])%>%as.numeric%>%replace_na(0),\n",
    "        cs = map_chr(susie_list$sets$cs[effect_id],~susie_tb$variants[.x]%>%paste0(collapse = \";\"))) ))%>%unnest(result_tb)\n",
    "        susie_effect_tb_ls$cs[susie_effect_tb_ls$cs == \"\"] = \"None\"\n",
    "        susie_effect_tb_ls %>% write_delim(paste0(\"${_output[0]:nnnn}.\",susie_list$name,\".effect.tsv\"),\"\\t\") \n",
    "    } \n",
    "  \n",
    "bash: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container\n",
    "    for i in ${_output:nr}; do\n",
    "    bgzip $i;\n",
    "    done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.23.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
