{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "76fc286c-f768-475f-9d7a-ab51056cd640",
   "metadata": {},
   "source": [
    "\n",
    "`\r",
    "## MWE\n",
    "\n",
    "\n",
    "\r\n",
    "\n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3281c730-a597-4cf6-bff5-ab5fc2a18cc3",
   "metadata": {},
   "source": [
    "# Command spliter\n",
    "This utility module is used to split a series of sos command into different genes to accomodate the need of AWS.\n",
    "## Input\n",
    "1. sos_command_file: A text file containing variouse sos command in the correct order. The output of command generator for 1 analysis is a suitable input as long as the order of operation are correct. It is assums that all the input sos command has a parameter called --region-name and the analysis can be partitioned by this parameter.\n",
    "2. region_list: A table where the last column is the name of the region to be split analysis.\n",
    "4. virtual_machine_path: The path of the file and working directories in the to be established virtual machine.\n",
    "5. wildcard_file (WIP, please ignore)\n",
    "6. output_file: The surfix of the output script.\n",
    "7. aws_command_file: A text file containing the command template for mm_jobman.sh, see https://github.com/rfeng2023/mmcloud/pull/13\n",
    "8. genotype_list: list of file for the genotype file per chrom, regardless of what path you put in the genotype_chrom_list, only the actual file name will be used. The relative path are to be specified in the vertual machine path  parameter.\r\n",
    "\n",
    "\n",
    "## Output\n",
    "1. a set of all the sos run command in the sos command_file, in the forms of for each regions in the region list within each chrom. Each row is a region. Each file is a chromosome. The `--genoFile {} --` patteren within each command will be replaced with the genotype file of that specific chromosome per the instruction of the genotype file\n",
    "   \n",
    "    `{command 1} --region-name {gene_name} && {command 2} --region-name {gene_name}`\n",
    "## MWE\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b97775c-c9ec-47c1-9c88-8e6430dfd9a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sos run command_spliter.ipynb split_command --sos_command_file test_sos_command --aws_command_file test_aws_command \\\n",
    "    --genotype_list /sc/arion/projects/CommonMind/roussp01a/snmulti_QTL/output/Genotype/snmulti_QTL_analysis_psychAD_ADSP.leftnorm.bcftools_qc.plink_qc.genotype_by_chrom_files.txt \\\n",
    "    --virtual_machine_path \"analysis/\" \\\n",
    "    --region_list  test_command_region_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34ed79ae-ecf1-440b-8d22-9d9cef6647f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "[global]\n",
    "parameter: sos_command_file = path\n",
    "parameter: aws_command_file = path\n",
    "parameter: wildcard_file = path(\"\")\n",
    "#parameter: s3_path = str\n",
    "parameter: virtual_machine_path = str\n",
    "parameter: region_list = path\n",
    "parameter: cwd = path(\".\")\n",
    "parameter: output_file = path(\"splited_cripts.sh\")\n",
    "parameter: genotype_list = path\n",
    "import pandas as pd\n",
    "chrom = list(set(pd.read_csv(region_list,sep =\"\\t\")[\"#chr\"]))\n",
    "\n",
    "[split_command_1]\n",
    "input: sos_command_file, for_each = \"chrom\"\n",
    "output: f'{cwd}/{_input:bn}.{_chrom}.sos_splited_script'\n",
    "\n",
    "import pandas as pd\n",
    "import re\n",
    "region = pd.read_csv(region_list,sep = \"\\t\")\n",
    "region = region[region.iloc[:, 0] == _chrom].iloc[:,-1].tolist() ## Subset the region\n",
    "genotype_list = pd.read_csv(genotype_list,sep = \"\\t\")\n",
    "per_chrom_genoFile = genotype_list [genotype_list.iloc[:, 0].astype(\"str\").replace(\"chr\",\"\") == str(_chrom).replace(\"chr\",\"\")].iloc[0,-1]\n",
    "per_chrom_genoFile = per_chrom_genoFile.split(\"/\")[-1]\n",
    "\n",
    "# Step 1: Read and modify the commands from a text file\n",
    "commands = []\n",
    "with open(_input, 'r') as file:\n",
    "    command = \"\"\n",
    "    for line in file:\n",
    "        stripped_line = line.strip()\n",
    "        if stripped_line.startswith('sos run'):\n",
    "            if command:\n",
    "                # Replace text between --genoFile and -- using regular expression\n",
    "                command = re.sub(r'(--genoFile ).*?( --)', r'\\1' + f'{virtual_machine_path}/{per_chrom_genoFile}' + r'\\2', command)\n",
    "                commands.append(command)\n",
    "            command = stripped_line\n",
    "        elif command.endswith('\\\\'):\n",
    "            command = command[:-1].strip() + ' ' + stripped_line\n",
    "    if command:\n",
    "        # Replace text in the last command\n",
    "        command = re.sub(r'(--genoFile ).*?( --)', r'\\1'  + f'{virtual_machine_path}/{per_chrom_genoFile}'  + r'\\2', command)\n",
    "        commands.append(command)\n",
    "\n",
    "# Step 2: Create a DataFrame with region names appended to each command\n",
    "data = {command: [f\"{command} --region-name {r}\" for r in region] for command in commands}\n",
    "df_commands = pd.DataFrame(data, index=region)\n",
    "\n",
    "# Step 3: Creating the concatenated list\n",
    "concatenated_commands = [\" && \".join(row) for _, row in df_commands.iterrows()]\n",
    "\n",
    "# Define your paths for the copy command\n",
    "path2 = virtual_machine_path\n",
    "\n",
    "# Step 4: Modifying each command\n",
    "modified_commands = [f\" {cmd} \" for cmd in concatenated_commands]\n",
    "\n",
    "output_file = f'{modified_commands}.sh'\n",
    "with open(_output, 'w') as f:\n",
    "    for cmd in modified_commands:\n",
    "        f.write(cmd + '\\n')\n",
    "\n",
    "[split_command_2]\n",
    "input: group_with = \"chrom\"\n",
    "output: f'{cwd}/{_input:bn}.jobman.sh'\n",
    "genotype_list = pd.read_csv(genotype_list,sep = \"\\t\")\n",
    "per_chrom_genoFile = genotype_list[genotype_list.iloc[:, 0].astype(\"str\").replace(\"chr\",\"\") == str(_chrom).replace(\"chr\",\"\")].iloc[0,-1]\n",
    "per_chrom_genoFile = per_chrom_genoFile.split(\"/\")[-1].rsplit('.', 1)[0]\n",
    "\n",
    "with open(aws_command_file, 'r') as file:\n",
    "    lines = file.readlines()\n",
    "    \n",
    "    # Identify the index of the second non-empty line\n",
    "second_non_empty_line_index = [i for i, line in enumerate(lines) if line.strip()][2]\n",
    "\n",
    "\n",
    "# Modify the second non-empty line\n",
    "lines[second_non_empty_line_index] = f\"./{_input} \\\\\\n\"\n",
    "\n",
    "# Append to the last line\n",
    "lines[-1] = lines[-1].strip() + f'  --download-include {per_chrom_genoFile}.bed {per_chrom_genoFile}.bim {per_chrom_genoFile}.fam \\n'\n",
    "\n",
    "# Write the modified content to a new file\n",
    "with open(_output, 'w') as file:\n",
    "    file.writelines(lines)\n",
    "\n",
    "[split_command_3]\n",
    "input: group_by = \"all\"\n",
    "output: f'{cwd}/{sos_command_file:bn}.{aws_command_file:bn}.all_chrom.jobman.sh'\n",
    "# List of file paths to concatenate\n",
    "# Open the new file in write mode\n",
    "with open(_output, 'w') as new_file:\n",
    "    # Iterate through each file in the list\n",
    "    for file_path in _input:\n",
    "        # Open and read each file\n",
    "        with open(file_path, 'r') as file:\n",
    "            # Write the content to the new file\n",
    "            new_file.write(file.read())\n",
    "            # Optionally add a newline between files\n",
    "            new_file.write('\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1b0860f-8b5c-4631-ba07-1fcbf3e9b3fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c59cabd-4896-433f-8796-e1ad5bc0fd3a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
